### ISSUE TITLE
[BUG] Incorrect Posterior Probability with dnUniformInteger and mvRandomIntegerWalk

### ISSUE THREAD

**Issue by minghao-du** [revbayes/revbayes#435]
**Describe the bug**
When using `dnUniformInteger` and `mvRandomIntegerWalk`, it is not possible to obtain the correct posterior probability.

**To Reproduce**
I generated 1000 samples ($y_i, i=1,...,1000$) from a Bernoulli distribution with $p=0.1$ and conducted inference on $p$.
$$y_i \sim \text{Bernoulli}(p)$$

$$p_\text{int} \sim \text{UniformInteger}(0,10)$$

$$p \coloneqq \frac{p_\text{int}}{10}$$

However, the inferred posterior distribution of $p_\text{int}$ is almost the same as the prior, uniformly distributed between $0$ and $10$, rather than being mostly $1$.
<img width="1171" alt="image" src="https://github.com/revbayes/revbayes/assets/51898298/6465ee51-f366-44ec-8c02-2f65377e5eee">

When inferring directly using a uniform distribution, there is no problem with the inference.
$$y_i \sim \text{Bernoulli}(p)$$

$$p_\text{int} \sim \text{Uniform}(0,10)$$

$$p \coloneqq \frac{p_\text{int}}{10}$$
<img width="1180" alt="image" src="https://github.com/revbayes/revbayes/assets/51898298/e8a43498-bb23-48a9-88c3-8e99d93d2100">

The scripts used and the output are in this file.
[UniInt_bug.zip](https://github.com/revbayes/revbayes/files/14787938/UniInt_bug.zip)

**Expected behavior**
When inferring using a UniformInteger distribution, the inferred $p_\text{int}$ is $1$.

**Computer info**
I am using a Mac, and the version of RevBayes I am using is development (rapture-2388-ge4baa8).

**Comment by bredelings** [comment]
Note that the Posterior and Likelihood are constant.  So they aren't getting updated when p_int changes.  This is the underlying source of the problem.

I tweaked the script to replace the collection of Bernoulli observations with a single Binomial observation, and the problem doesn't go away.  So this could possibly be a problem with mvRandomIntegerWalk( )... but I'm not sure how the move can prevent the likelihood and prior from changing.

**Comment by bjoelle** [comment]
This seems tied to the `model()` call
 * when using `model(uni_int)` as in the example script, `p_estimated` and the Bernouilli distribution do not get integrated into the model (so `uni_int` updates but then nothing else does)
 * when using `model(p_estimated)` or `model(y)`, the graph does get included correctly, but `uni_int` is set as constant and loses its prior distribution (so `uni_int` doesn't update and neither does anything else)
 * when using `model(data)`, it gives the error "Cannot find node with name 'uni_int' in the model but received a move working on it." so I'm assuming it just lost `uni_int` entirely

Overall, there seems to be something weird going on with understanding the graph, tied to (I assume) **deterministic nodes** and/or **integer parameters**

**Comment by bredelings** [comment]
Interestingly, it looks like this problem seems not to happen if the integer distribution is `dnPoisson( )` instead of `dnUniformInteger( )`.

**Comment by bredelings** [comment]
Here's what we see if dnPoisson is used (and the sample size is changed to 10:
![Screenshot from 2024-04-09 16-52-22](https://github.com/revbayes/revbayes/assets/140268/b9275dde-5936-48a4-80bc-2d2f300a5d3c)

**Comment by bredelings** [comment]
Here is what we see if dnUniformInteger is used:
![Screenshot from 2024-04-09 16-53-30](https://github.com/revbayes/revbayes/assets/140268/a845b4cd-93c5-46f0-9431-50711a949a32)

**Comment by bjoelle** [comment]
I've looked into this further, and it seems to be tied to the use of arithmetic functions with an implicit conversion
using the script here 
```uni_int ~ dnUniformInteger(0, 10)
uni_x := uni_int / 10
uni_t := abs(uni_int)
structure(uni_x)
structure(uni_int)
structure(uni_t)
```
we can see that `uni_x` and `uni_t` get created properly with `uni_int` as parent, but `uni_int` does not register them correctly as children
the exact same code but with `uni_int ~ dnUniform(0, 10)` correctly links the elements both ways

**Comment by bjoelle** [comment]
so as far as I can tell the conversion goes through `Integer::convertTo` which calls `ModelObject<rbType>::ModelObject(rbType *v)` and this appears to create a constant node no matter what (which is a bit weird for a conversion, shouldn't that be a deterministic node ?)
anyway the final result is you end up with one `uni_int` which is an int and stochastic, and which is the workspace version (aka the one that gets printed by structure) and **another** `uni_int` which is a real and constant, and which is the one referenced by `uni_x`
thus the mess

**Comment by bredelings** [comment]
Hmm... I think there has been a long-standing issue where revbayes converts between types without using a deterministic node.

Suppose we have `f(g(x))` where `g` produces an Integer, but `f` wants a Real.  What revbayes does is to change the type of the `g(x)` deterministic node to Real when `f` uses it.  But if `g(x)` is updated, then it will change back to Integer.  So the type of `g(x)` alternative Integer, Real, Integer, Real, etc...

The only reason for this is that it is supposed to be faster than producing a deterministic node for the type conversion.  But its not clear that producing a deterministic node would actually be slower....

One thing we could try would simply be to make type conversions rely on a deterministic node.  We could see if it is actually slower.  Then if it is not slower, we could the situation where the type of a node alternatives between 2 types forever...

**Comment by bjoelle** [comment]
so here the situation doesn't seem to be that the type alternate, but rather than two nodes of each type get created and the wrong one is referenced
possibly this is because the Integer node already has a name here, unlike your example

**Comment by bjoelle** [comment]
I think having conversions as deterministic nodes (at least if the converted node is not constant) makes sense, and I don't see a way to avoid it if we want to convert variables which can be updated after conversion

**Comment by bjoelle** [comment]
should be resolved following PR #545
